src/mod_offline_riak.erl

```diff
@@ -184,3 +185,37 @@ maybe_decode_timestamp(?INFINITY) ->
 maybe_decode_timestamp(TS) ->
     usec:to_now(TS).
 
+
+get_personal_data(Username, Server) ->
+    LServer = jid:nodeprep(Server),
+    [{offline, ["timestamp", "from", "to", "packet"], fetch_messages(Username, LServer)}].
+
+fetch_messages(Username, LServer) ->
+    LUser = jid:nodeprep(Username),
+    Keys = read_user_idx(LUser, LServer),
+    User = jid:to_binary({Username, LServer}),
+    Msgs = [fetch_msg(Key, LServer, User) || Key <- Keys],
+    lists:flatten(Msgs).
+
+fetch_msg(Key, LServer, To) ->
+    try
+        {ok, Obj} = mongoose_riak:get(bucket_type(LServer), Key),
+
+        PacketRaw = riakc_obj:get_value(Obj),
+        {ok, Packet} = exml:parse(PacketRaw),
+        MD = riakc_obj:get_update_metadata(Obj),
+        [Timestamp] = riakc_obj:get_secondary_index(MD, ?TIMESTAMP_IDX),
+        From = riakc_obj:get_user_metadata_entry(MD, <<"from">>),
+
+        NowUniversal = calendar:now_to_universal_time(usec:to_now(Timestamp)),
+        {UTCTime, UTCDiff} = jlib:timestamp_to_iso(NowUniversal, utc),
+        UTC = list_to_binary(UTCTime ++ UTCDiff),
+        {UTC, jid:to_binary(jid:binary_to_bare(From)), To, exml:to_binary(Packet)}
```

Comment:
Username is provided as `To`, not JID.

Reply:
I think everything is fine, just again the variable name is confusing